rhino/src/main/java/org/mozilla/javascript/regexp/UnicodeProperties.java

```diff
@@ -0,0 +1,437 @@
+                    try {
+                        return encodeProperty(
+                                SCRIPT, (byte) Character.UnicodeScript.forName(value).ordinal());
+                    } catch (IllegalArgumentException e) {
+                        return -1;
+                    }
+                default:
+                    // Binary properties don't have values
+                    return -1;
+            }
+        }
+    }
+
+    /**
+     * Encodes a property name and value into a single int. The property name is in the high 16
+     * bits, the value in the low 16 bits.
+     *
+     * @param property Property name constant
+     * @param value Property value constant
+     * @return Encoded int
+     */
+    private static int encodeProperty(byte property, byte value) {
+        return ((property & 0xFF) << 8) | (value & 0xFF);
+    }
+
+    private static final Character.UnicodeScript[] UnicodeScriptValues =
+            Character.UnicodeScript.values();
+
+    /**
+     * Tests if a code point has a specific Unicode property.
+     *
+     * @param property Encoded property (from lookup method)
+     * @param codePoint Character code point to test
+     * @return true if the code point has the property
+     */
+    public static boolean hasProperty(int property, int codePoint) {
+        byte propByte = (byte) ((property >> 8) & 0xFF);
+        int valueByte = (property & 0xFF);
+
+        switch (propByte) {
+            case ALPHABETIC:
+                return Character.isAlphabetic(codePoint) == (valueByte == TRUE);
+
+            case ASCII:
+                return (codePoint <= 0x7F) == (valueByte == TRUE);
+
+            case CASE_IGNORABLE:
+                // Java doesn't have a direct method for this
+                // This is an approximation
+                return (Character.getType(codePoint) == Character.MODIFIER_SYMBOL
+                                || Character.getType(codePoint) == Character.MODIFIER_LETTER
+                                || Character.getType(codePoint) == Character.NON_SPACING_MARK)
+                        == (valueByte == TRUE);
+
+            case GENERAL_CATEGORY:
+                int javaCategory = Character.getType(codePoint);
+                return checkGeneralCategory(valueByte, javaCategory);
+            case HEX_DIGIT:
+                return isHexDigit(codePoint) == (valueByte == TRUE);
+
+            case ID_CONTINUE:
+                return Character.isUnicodeIdentifierPart(codePoint) == (valueByte == TRUE);
+
+            case ID_START:
+                return Character.isUnicodeIdentifierStart(codePoint) == (valueByte == TRUE);
+
+            case LOWERCASE:
+                return Character.isLowerCase(codePoint) == (valueByte == TRUE);
+
+            case UPPERCASE:
+                return Character.isUpperCase(codePoint) == (valueByte == TRUE);
+
+            case WHITE_SPACE:
+                return Character.isWhitespace(codePoint) == (valueByte == TRUE);
```

Comment:
Unicode's notion of whitespace maps to `Character.isSpaceChar`, not `isWhiteSpace`, because of different handling of control characters and non-breaking spaces.

`"\u00A0".match(/\p{space}/u)`